File: calc_vel.py
import numpy as np
from Get_global_value import num_q
from Get_global_value import J_type
from Get_global_value import Ez
from Get_global_value import BB
from Get_global_value import m0
from Get_global_value import m
from Get_global_value import mass
from Get_global_value import inertia0
from Get_global_value import inertia
from Get_global_value import cc
from calc_jr import calc_jr
from calc_jt import calc_jt
from cross import cross
from Get_global_value import c0


def calc_vel(A0, AA, v0, w0, q, qd):
    vv = np.zeros((num_q, 3))
    ww = np.zeros((num_q, 3))
    if num_q == 0:
        print('Single body, there is no link')
    else:
        for i in range(num_q):
            if BB[i] == -1:
                A_I_i = AA[i, :, :]
                if J_type[i] == 'R':
                    ww[i, :] = w0[0:3] + A_I_i @ Ez * qd[i]
                    vv[i, :] = v0[0:3] + cross(w0[0:3], A0 @ c0[i, :]) - cross(ww[i, :], A_I_i @ cc[i, i, :])
                else:
                    ww[i, :] = w0[0:3]
                    vv[i, :] = v0[0:3] \
                               + cross(w0[0:3], np.dot(A0, c0[i, :])) \
                               - cross(ww[i, :], np.dot(A_I_i, cc[i, i, :])) \
                               + cross(ww[i, :], np.dot(np.dot(A_I_i, Ez), q[i])) \
                               + np.dot(np.dot(A_I_i, Ez), qd[i])

            else:
                A_I_BB = AA[BB[i], :, :]
                A_I_i = AA[i, :, :]
                if J_type == 'R':
                    ww[i, :] = ww[BB[i], :] + np.dot(np.dot(A_I_i, Ez), qd[i])
                    vv[i, :] = vv[BB[i], :] \
                               + cross(ww[BB[i], :], np.dot(A_I_BB, cc[BB[i], i, :])) \
                               - cross(ww[i, :], np.dot(A_I_i, cc[i, i, :]))
                else:
                    ww[i, :] = ww[BB[i], :]
                    vv[i, :] = vv[BB[i], :] \
                               + cross(ww[BB[i], :], np.dot(A_I_BB, cc[BB[i], i, :])) \
                               - cross(ww[i, :], np.dot(A_I_i, cc[i, i, :])) \
                               + cross(ww[i, :], np.dot(np.dot(A_I_i, Ez), q[i])) \
                               + np.dot(np.dot(A_I_i, Ez), qd[i])

    return vv, ww

File: src/matlab/spacedyn_v2r1/calc_vel.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% The SpaceDyn, a MATLAB toolbox for Space and Mobile Robots.
% (C)1998 The Space Robotics Lab. directed by Kazuya Yoshida,
% Tohoku University, Japan.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%	February 4, 1999, Last modification by K.Yoshida
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%
% CALC_VEL	Calculate the velocity of each link
%
%   CALC_VEL returns the link velocity in the inertial frame
%            for link 1 to n.
%
%   1997 6/12 T.Hiraoka
%   2001.9.17 H.Hamano
%   2002.2.28 A.Noguchi modified to version up
%

function SV = calc_vel( LP, SV )

%global BB J_type
%global c0 cc
%global num_q Ez
global Ez

% If a Single body,
if LP.num_q == 0
   
   SV.vv = [];
   SV.ww = [];
   
% If a Multi body system,
else
   
   % Calculation of velocity vectors vv,ww
   for i = 1 : LP.num_q
      
      % Check the link connection: Is the lower one of this link, 0 ?
      if LP.BB(i) == 0
         
         % Current (i-th) link connects to the 0-th link
         A_I_i = SV.AA(:,i*3-2:i*3);
         
         % Rotational joint
         if LP.J_type(i) == 'R'
            SV.ww(:,i) = SV.w0(:) + A_I_i*Ez*SV.qd(i);
            SV.vv(:,i) = SV.v0(:) ...
               + cross( SV.w0(:),(SV.A0*LP.c0(:,i)) ) ...
               - cross( SV.ww(:,i),(A_I_i*LP.cc(:,i,i)) );
            
         % Prismatic joint
         else
            SV.ww(:,i) = SV.w0(:);
            SV.vv(:,i) = SV.v0(:) ...
               + cross( SV.w0(:),(SV.A0*LP.c0(:,i)) ) ...
               - cross( SV.ww(:,i),(A_I_i*LP.cc(:,i,i)) ) ...
               + cross( SV.ww(:,i),(A_I_i*Ez*SV.q(i)) ) ...
               + A_I_i*Ez*SV.qd(i);
            
         end
         
      % Current (i-th) link doesn't connect to the 0-th link
      else
         
         A_I_BB = SV.AA(:,LP.BB(i)*3-2:LP.BB(i)*3);
         A_I_i  = SV.AA(:,i*3-2:i*3);
         
         % Rotational joint
         if LP.J_type(i) == 'R'
            SV.ww(:,i) = SV.ww(:,LP.BB(i)) + A_I_i*Ez*SV.qd(i);
            SV.vv(:,i) = SV.vv(:,LP.BB(i)) ...
               + cross( SV.ww(:,LP.BB(i)),(A_I_BB*LP.cc(:,LP.BB(i),i)) ) ...
               - cross( SV.ww(:,i),(A_I_i*LP.cc(:,i,i)) );
            
         % Prismatic joint
         else
            SV.ww(:,i) = SV.ww(:,LP.BB(i));
            SV.vv(:,i) = SV.vv(:,LP.BB(i)) ...
               + cross( SV.ww(:,LP.BB(i)),(A_I_BB*LP.cc(:,LP.BB(i),i)) ) ...
               - cross( SV.ww(:,i),(A_I_i*LP.cc(:,i,i)) ) ...
               + cross( SV.ww(:,i),(A_I_i*Ez*SV.q(i)) ) ...
               + A_I_i*Ez*SV.qd(i);
            
         end
         
      end
      
   end
   
end


%%%EOF
